### PR TITLE
Rarest First Algorithm

### PR DESCRIPTION
**Pull request by ketphan02**
### Sample Output:
```
Class size: 8
[4, 5]
Class size: 20
[0, 15]
Class size: 60
[26, 39]
Class size: 100
[57, 12]
Class size: 200
[0, 147]
```

### CODE REVIEW

File: api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+"""
+Rarest First Algorithm
+
+TODO: Write up the algorithm steps here
```
Comment (ketphan02) [inline api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py:4]: Finish TODO

File: api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            ].get(attribute_id)
+            if distance_from_i_star and distance_from_i_star.is_updated:
+                result.add(distance_from_i_star.end_student.id)
+
+        team = Team(
+            name="Subgraph",
+            students=list(student_id_trace.get(student_id) for student_id in result),
+            _id=0,
+        )
+        return TeamSet(teams=[team])
```
Comment (Opeyem1a) [inline api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py:169]: why only 1 team?

File: api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+        result: Set[int] = {min_distance.start_student.id}
+        for attribute_id in self.attributes:
+            distance_from_i_star = shortest_distance_to_attribute[
+                min_distance.start_student.id
+            ].get(attribute_id)
+            if distance_from_i_star and distance_from_i_star.is_updated:
+                result.add(distance_from_i_star.end_student.id)
+
+        team = Team(
```
Comment (Opeyem1a) [inline api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py:164]: see my comment on your other PR about `self.teams`

**Review by Opeyem1a — COMMENTED** [review]
Seems fine besides small comment

File: api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py (hunk 2 above)
Comment (ketphan02) [reply api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py:169]: This algorithm only outputs 1 team (sub graph in a social network)

**Review by Opeyem1a — COMMENTED** [review]
probably mergeable (?)

**Review by Opeyem1a — APPROVED** [review]

**Pull request closed without merging**